include/deal.II/matrix_free/operators.h

```diff
@@ -35,6 +35,345 @@ DEAL_II_NAMESPACE_OPEN
+      for (unsigned int i=0; i<edge_constrained_indices[0].size(); ++i)
+        dst.local_element(edge_constrained_indices[0][i]) = 1.;
+    }
+
+
+
+    template <int dim, typename VectorType>
+    typename std_cxx11::enable_if<IsBlockVector<VectorType>::value, void>::type
+    external_set_constrained_entries_to_one
+    (const MatrixFree<dim, typename VectorType::value_type> &mf,
+     const std::vector<std::vector<unsigned int> > &edge_constrained_indices,
+     VectorType &dst)
+    {
+      AssertDimension(edge_constrained_indices.size(), dst.n_blocks());
+      for (unsigned int j = 0; j < dst.n_blocks; ++j)
+        {
+          const std::vector<unsigned int> &constrained_dofs = mf.get_constrained_dofs(j);
+          for (unsigned int i=0; i<constrained_dofs.size(); ++i)
+            dst.block(j).local_element(constrained_dofs[i]) = 1.;
+          for (unsigned int i=0; i<edge_constrained_indices[j].size(); ++i)
+            dst.block(j).local_element(edge_constrained_indices[j][i]) = 1.;
+        }
+    }
+
+
+    template <typename VectorType>
+    typename std_cxx11::enable_if<IsBlockVector<VectorType>::value, void>::type
+    external_set_dirichlet_zero
+    (std::vector<std::vector<std::pair<typename VectorType::value_type, typename VectorType::value_type> >> &edge_constrained_values,
+     const std::vector<std::vector<unsigned int> > &edge_constrained_indices,
+     VectorType &src,
+     const VectorType &dst)
+    {
+      typedef typename VectorType::value_type Number;
+      AssertDimension(src.n_blocks(), src.n_blocks());
+      AssertDimension(edge_constrained_values.size(), src.n_blocks());
+      AssertDimension(edge_constrained_indices.size(), src.n_blocks());
+      for (unsigned int i = 0; i < src.n_blocks(); ++i)
+        {
+          for (unsigned int j = 0; j < edge_constrained_indices[i].size(); ++j)
+            {
+              edge_constrained_values[i][j] = std::pair<Number, Number>
+                                              (src.block(i).local_element(edge_constrained_indices[i][j]),
+                                               dst.block(i).local_element(edge_constrained_indices[i][j]));
+              src.block(i).local_element (edge_constrained_indices[i][j]) = 0.;
+            }
+        }
+
+    }
+
+
+
+    template <typename VectorType>
+    typename std_cxx11::enable_if<!IsBlockVector<VectorType>::value, void>::type
+    external_set_dirichlet_zero
+    (std::vector<std::vector<std::pair<typename VectorType::value_type,
+     typename VectorType::value_type> >> &edge_constrained_values,
+     const std::vector<std::vector<unsigned int> > &edge_constrained_indices,
+     VectorType &src,
+     const VectorType &dst)
+    {
+      typedef typename VectorType::value_type Number;
+      AssertDimension(edge_constrained_values.size(), 1);
+      AssertDimension(edge_constrained_indices.size(), 1);
+      for (unsigned int j = 0; j < edge_constrained_indices[0].size(); ++j)
```

Comment:
In the function above, I access a block of the vector which I can't for non-block vectors. Therefore, I don't immediately see how to avoid splitting this function.